Add in-depth introduction for Fortran with make (#160)
fortran-lang/fortran-lang.org: learn/building_programs/project_make.md

---

We have shortly discussed the basics of ``make``, this chapter gives thoughts
and strategies to make ``make`` scale for larger projects.

Before going into detail with ``make``, there are a few points to consider first:
1. ``make`` is a Unix tool and might give you a hard time when porting to non-Unix
   platforms. That said, there are also different flavors of ``make`` available,
   not all might support the features you want to use.
2. while ``make`` gives you full control over the build process, it also
   means you are fully responsible for the entire build process, from the start
   to the end in every detail you have to specify the rules for your project.
   You might find yourself spending a significant amount of time writing and
   maintaining your ``Makefile``, this is usually time you cannot use for
   developing your source.
3. you can work with your ``Makefile``, but think about the other developers
   on your project (your friends or your coworkers or maybe somebody visiting your
   open source project). How much time do you expect them to spend learning your
   ``Makefile`` and would they be able to debug or add functionalities?

> 4. pure ``make`` will not scale. You will soon write or use auxiliary programs
>    to dynamically or statically generate your ``Makefile``, those introduce
>    dependencies and possible sources of errors. Testing and documenting those
>    tools should not be neglected.

Reviewer: ```suggestion
4. Pure ``make`` will not scale. You will soon add auxiliary programs
   to dynamically or statically generate your ``Makefile``. Those introduce
   dependencies and possible sources of errors. The effort needed to test and document those
   tools should not be underestimated.
```

If you think ``make`` is suitable for your needs, than you can start writing
your ``Makefile``. For this course we will use real world examples from the

> package index, which use (at the point of writing) different build systems

Reviewer: ```suggestion
package index, which (at the time of writing) use build systems other
```

than ``make``. This guide should present a general recommended style to write
``make``, but also serve as demonstration of useful and interesting features.

{% include tip.html content="Even if you find ``make`` unsuitable to build your project, it is *the* tool to automate workflows defined by files. Maybe you can leverage its power in a different context." %}

## Getting started

For this part we will work with
<a href="https://github.com/jacobwilliams/fortran-csv-module/tree/1.2.0" target="_blank" rel="noopener"> the Fortran CSV module (v1.2.0)</a>.

> Aim is it to write a ``Makefile`` to compile this project to a static library.

Reviewer: ```suggestion
Our goal is to write a ``Makefile`` to compile this project to a static library.
```

Start by cloning the repository

```
git clone https://github.com/jacobwilliams/fortran-csv-module -b 1.2.0
cd fortran-csv-module
```

{% include note.html content="For this part we will work with the code from tag ``1.2.0``, to make it as reproducible as possible. Feel free to use the latest version or another project instead." %}

> The project just downloaded uses FoBiS as build system, you can check the

Reviewer: ```suggestion
This project uses FoBiS as build system, and you can check the
```

``build.sh`` for options used with FoBiS. We are about to write a ``Makefile``
for this project. First, we check the directory structure and the source files

    .
    ├── build.sh
    ├── files
    │   ├── test_2_columns.csv
    │   └── test.csv
    ├── fortran-csv-module.md
    ├── LICENSE
    ├── README.md
    └── src
        ├── csv_kinds.f90
        ├── csv_module.F90
        ├── csv_parameters.f90
        ├── csv_utilities.f90
        └── tests
            ├── csv_read_test.f90
            ├── csv_test.f90
            └── csv_write_test.f90

> We find seven differently named Fortran source files, the four in ``src`` should

Reviewer: ```suggestion
We find seven different Fortran source files; the four in ``src`` should
```

be compiled and added to a static library while the three in ``src/tests``

> contain individual programs depending on this static library.

Reviewer: ```suggestion
contain individual programs that depend on this static library.
```

    ar rcs libcsv.a src/csv_kinds.f90.o src/csv_module.F90.o src/csv_parameters.f90.o src/csv_utilities.f90.o
    gfortran -c -o src/tests/csv_read_test.f90.o src/tests/csv_read_test.f90
    gfortran -o src/tests/csv_read_test.exe src/tests/csv_read_test.f90.o libcsv.a
    gfortran -c -o src/tests/csv_test.f90.o src/tests/csv_test.f90
    gfortran -o src/tests/csv_test.exe src/tests/csv_test.f90.o libcsv.a
    gfortran -c -o src/tests/csv_write_test.f90.o src/tests/csv_write_test.f90
    gfortran -o src/tests/csv_write_test.exe src/tests/csv_write_test.f90.o libcsv.a

There are a few things to note there, a ``make`` build usually interlaces the
build artifacts and the source code, unless you put extra effort into implementing
a build directory.
Also, right now the the source files and dependencies are specified explicitly,
which results in several additional lines even for such a simple project.

## Automatically generated dependencies

The main drawback of ``make`` for Fortran is the missing capability to
determine module dependencies. This is usually solved by either adding those
by hand or automatically scanning the source code with an external tool.

> Some compilers also offer to generate dependencies in ``make`` format.

Reviewer: I didn't know this. Which compilers can do this?

Perhaps we can add a few examples in parentheses, i.e. "Some compilers (e.g. this compiler or that compiler)"

Author: `ifort -gen-deps` is one example, about I never used it so far.

Author: ```suggestion
Some compilers (like the Intel Fortran compiler) also offer to generate dependencies in ``make`` format.
```

Before diving into the dependency generation, we will outline the concept of
a robust take on the dependency problem.
First, we want an approach that can process all source files independently,
while each source file provides (``module``) or requires (``use``) modules.
When generating the dependencies only the name of the source file and the

> module files are known, no information on the object file names should be

Reviewer: ```suggestion
module files are known, and no information on the object file names should be
```

```

Note the declaration of ``obj`` as recursively expanded variable, we effectively
use this mechanism to define a function in ``make``. The ``foreach`` function
allows us to loop over all source files, while the ``eval`` function allows us
to generate ``make`` statements and evaluate them for this ``Makefile``.

We adjust the dependencies accordingly as we can now define the name of the
object files through the source file names:

```make
# Define all module interdependencies
csv_kinds.mod := $(src/csv_kinds.f90)
csv_module.mod := $(src/csv_module.F90)
csv_parameters.mod := $(src/csv_parameters.f90)
csv_utilities.mod := $(src/csv_utilities.f90)
$(src/csv_module.F90): $(csv_utilities.mod)
$(src/csv_module.F90): $(csv_kinds.mod)
$(src/csv_module.F90): $(csv_parameters.mod)
$(src/csv_parameters.f90): $(csv_kinds.mod)
$(src/csv_utilities.f90): $(csv_kinds.mod)
$(src/csv_utilities.f90): $(csv_parameters.mod)
$(src/tests/csv_read_test.f90): $(csv_module.mod)
$(src/tests/csv_test.f90): $(csv_module.mod)
$(src/tests/csv_write_test.f90): $(csv_module.mod)
```

The same strategy of creating a map is already used for the module files, now
it is just expanded to the object files as well.

> To generate the respective dependency map automatically we will use a simple

Reviewer: ```suggestion
To generate the respective dependency map automatically we will use an
```

For somebody not familiar with awk, this script is anything but simple. I'd even say it looks daunting. For such a reader it could be discouraging to read the words "simple script" and then see the thing that follows.

``awk`` script here

```awk
#!/usr/bin/awk -f

BEGIN {
    IGNORECASE = 1
}

$1 ~ /^module$/ &&
$2 ~ /^[a-zA-Z][a-zA-Z1-9_]*$/ {
    if (modc[FILENAME,$2]++ == 0) {
        mod[++im] = sprintf("%s.mod = $(%s)", tolower($2), FILENAME)
    }
}

$1 ~ /^use$/ &&
$2 ~ /^[a-zA-Z][a-zA-Z1-9_]*,?$/ {
    gsub(/,/, "", $2)
    if (usec[FILENAME,$2]++ == 0) {
        use[++iu] = sprintf("$(%s) += $(%s.mod)", FILENAME, tolower($2))
    }
}

$1 ~ /^(#:?)?include$/ &&
$2 ~ /^["'].+["']$/ {
    gsub(/'|"/, "", $2)
    if (incc[FILENAME,$2]++ == 0) {
        inc[++ii] = sprintf("$(%s) += %s", FILENAME, $2)
    }
}

END {
    for (i in mod) print mod[i]
    for (i in use) print use[i]
    for (i in inc) print inc[i]
}
```

> This is a particular simple script which makes a few assumptions on the source
> code it is going to parse, so it will certainly not work for all Fortran code,
> but for this example it will suffice.

Reviewer: ```suggestion
This script makes a few assumptions about the source code it parses, so it will not work with all Fortran code, but for this example it will suffice.
```

``./gen-deps.awk $(find src -name '*.[fF]90')``. You should see output like this:

    csv_utilities.mod = $(src/csv_utilities.f90)
    csv_kinds.mod = $(src/csv_kinds.f90)
    csv_parameters.mod = $(src/csv_parameters.f90)
    csv_module.mod = $(src/csv_module.F90)
    $(src/csv_utilities.f90) += $(csv_kinds.mod)
    $(src/csv_utilities.f90) += $(csv_parameters.mod)
    $(src/csv_kinds.f90) += $(iso_fortran_env.mod)
    $(src/tests/csv_read_test.f90) += $(csv_module.mod)
    $(src/tests/csv_read_test.f90) += $(iso_fortran_env.mod)
    $(src/tests/csv_write_test.f90) += $(csv_module.mod)
    $(src/tests/csv_write_test.f90) += $(iso_fortran_env.mod)
    $(src/tests/csv_test.f90) += $(csv_module.mod)
    $(src/tests/csv_test.f90) += $(iso_fortran_env.mod)
    $(src/csv_parameters.f90) += $(csv_kinds.mod)
    $(src/csv_module.F90) += $(csv_utilities.mod)
    $(src/csv_module.F90) += $(csv_kinds.mod)
    $(src/csv_module.F90) += $(csv_parameters.mod)
    $(src/csv_module.F90) += $(iso_fortran_env.mod)

Note that the scripts output will use recursively expanded variables and not
define any dependencies yet, because out-of-order declaration of variables

> might be necessary and do not want to create any target by accident.

Reviewer: ```suggestion
might be necessary and we do not want to create any target by accident.
```
